substrate/frame/support/src/lib.rs

```diff
@@ -1064,1232 +961,941 @@ pub mod pallet_prelude {
+	/// }
+	/// ```
+	/// I.e. a regular enum named `Error`, with generic `T` and fieldless or multiple-field
+	/// variants.
+	///
+	/// Any field type in the enum variants must implement [`scale_info::TypeInfo`] in order to
+	/// be properly used in the metadata, and its encoded size should be as small as possible,
+	/// preferably 1 byte in size in order to reduce storage size. The error enum itself has an
+	/// absolute maximum encoded size specified by
+	/// [`frame_support::MAX_MODULE_ERROR_ENCODED_SIZE`].
+	///
+	/// (1 byte can still be 256 different errors. The more specific the error, the easier it
+	/// is to diagnose problems and give a better experience to the user. Don't skimp on having
+	/// lots of individual error conditions.)
+	///
+	/// Field types in enum variants must also implement [`frame_support::PalletError`],
+	/// otherwise the pallet will fail to compile. Rust primitive types have already
+	/// implemented the [`frame_support::PalletError`] trait along with some commonly used
+	/// stdlib types such as [`Option`] and [`sp_std::marker::PhantomData`], and hence
+	/// in most use cases, a manual implementation is not necessary and is discouraged.
+	///
+	/// The generic `T` must not bound anything and a `where` clause is not allowed. That said,
+	/// bounds and/or a where clause should not needed for any use-case.
+	///
+	/// ## Macro expansion
+	///
+	/// The macro implements the [`Debug`] trait and functions `as_u8` using variant position,
+	/// and `as_str` using variant doc.
+	///
+	/// The macro also implements `From<Error<T>>` for `&'static str` and `From<Error<T>>` for
+	/// `DispatchError`.
+	pub use frame_support_procedural::error;
+
+	/// Allows defining pallet events.
+	///
+	/// Pallet events are stored under the `system` / `events` key when the block is applied
+	/// (and then replaced when the next block writes it's events).
+	///
+	/// The Event enum can be defined as follows:
+	///
+	/// ```
+	/// #[frame_support::pallet(dev_mode)]
+	/// mod pallet {
+	/// #     use frame_support::pallet_prelude::IsType;
+	/// #
+	/// 	#[pallet::pallet]
+	/// 	pub struct Pallet<T>(_);
+	///
+	/// 	#[pallet::event]
+	/// 	#[pallet::generate_deposit(fn deposit_event)] // Optional
+	/// 	pub enum Event<T> {
+	/// 		/// SomeEvent doc
+	/// 		SomeEvent(u16, u32), // SomeEvent with two fields
+	/// 	}
+	///
+	/// 	#[pallet::config]
+	/// 	pub trait Config: frame_system::Config {
+	/// 		/// The overarching runtime event type.
+	/// 		type RuntimeEvent: From<Event<Self>>
+	/// 			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	/// 	}
+	/// }
+	/// ```
+	///
+	/// I.e. an enum (with named or unnamed fields variant), named `Event`, with generic: none
+	/// or `T` or `T: Config`, and optional w here clause.
+	///
+	/// `RuntimeEvent` must be defined in the `Config`, as shown in the example.
+	///
+	/// Each field must implement [`Clone`], [`Eq`], [`PartialEq`], [`codec::Encode`],
```

Comment:
nit: We could probably mention `TypeInfo` similar to errors, to ensure that it is collected by metadata?